Type checking error in patmat (related to overloaded unapply)

Following code compiles fine in scalac:
```Scala
object Test {
  class S(i: Int)
  case class T(i: Int) extends S(i)

  object T {
    def unapply(s: S): Option[(Int, Int)] = Some(5, 6)
  }

  val s = new S(5)

  s match {
    case T(x, y) => println(x + y)
    case _ => println("not match")
  }
}
```

But it fails with dotty:
```
examples/auto-tupling2.scala:13: error: Pattern type (T1, T2) is neither a subtype nor a supertype of selector type Int
    case T(x, y) => println(x + y)
           ^
examples/auto-tupling2.scala:13: error: type mismatch:
 found   : Nothing(x)
 required: ?{ + : ? }
Note that implicit conversions cannot be applied because they are ambiguous;
 both method Byte2byte in object Predef$ and method Character2char in object Predef$ convert from Nothing(x) to ?{ + : FunProto(y):Any }
    case T(x, y) => println(x + y)
                            ^
two errors found
```

Following change seems to fix the problem:

``` Scala
--- a/src/dotty/tools/dotc/typer/Applications.scala
+++ b/src/dotty/tools/dotc/typer/Applications.scala
@@ -709,11 +709,11 @@ trait Applications extends Compatibility { self: Typer =>
       // try first for non-overloaded, then for overloaded ocurrences
       def tryWithName(name: TermName)(fallBack: Tree => Tree)(implicit ctx: Context): Tree =
         tryEither {
-          implicit ctx => typedExpr(untpd.Select(qual, name), genericProto)
+          implicit ctx => typedExpr(untpd.Select(qual, name), specificProto)
         } {
           (sel, _) =>
             tryEither {
-              implicit ctx => typedExpr(untpd.Select(qual, name), specificProto)
+              implicit ctx => typedExpr(untpd.Select(qual, name), genericProto)
             } {
               (_, _) => fallBack(sel)
             }
```

What's the consideration to try the generic first, then the specific one?

--- a/src/dotty/tools/dotc/typer/Applications.scala
+++ b/src/dotty/tools/dotc/typer/Applications.scala
@@ -709,11 +709,11 @@ trait Applications extends Compatibility { self: Typer =>
       // try first for non-overloaded, then for overloaded ocurrences
       def tryWithName(name: TermName)(fallBack: Tree => Tree)(implicit ctx: Context): Tree =
         tryEither {
-          implicit ctx => typedExpr(untpd.Select(qual, name), genericProto)
+          implicit ctx => typedExpr(untpd.Select(qual, name), specificProto)
         } {
           (sel, _) =>
             tryEither {
-              implicit ctx => typedExpr(untpd.Select(qual, name), specificProto)
+              implicit ctx => typedExpr(untpd.Select(qual, name), genericProto)
             } {
               (_, _) => fallBack(sel)
             }

--- a/tests/pos/i1318.scala
+++ b/tests/pos/i1318.scala
@@ -1,0 +1,38 @@
+object Foo {
+  class S(i: Int)
+  case class T(i: Int) extends S(i)
+
+  object T {
+    def unapply(s: S): Option[(Int, Int)] = Some(5, 6)
+    // def unapply(o: Object): Option[(Int, Int, Int)] = Some(5, 6, 7)
+  }
+
+  val s = new S(5)
+
+  s match {
+    // case T(x, y, z) => println(x + y + z)
+    case T(x, y) => println(x + y)
+    case T(x) => println(x)
+    case _ => println("not match")
+  }
+}
+
+object Bar {
+  case class T(i: Int)
+  class S(i: Int) extends T(i)
+
+  object T {
+    def unapply(s: S): Option[(Int, Int)] = Some(5, 6)
+    // def unapply(o: Object): Option[(Int, Int, Int)] = Some(5, 6, 7)
+  }
+
+  val s = new S(5)
+
+  s match {
+    // case T(x, y, z) => println(x + y + z)
+    case T(x, y) => println(x + y)
+    case T(x) => println(x)
+    case _ => println("not match")
+  }
+}
+

